### ISSUE TITLE
improve diagnostic context for property vs regular attribute protocol assignability failure

### ISSUE THREAD

**Issue by alexprengere** [astral-sh/ty#3940]
### Summary

Hi! 
I just updated from 0.0.56 to 0.0.57, and noticed this error appearing. Here is a MRO:

```python
from typing import Generic, Protocol, TypeVar

class WithValue(Protocol):
    value: str | None

T = TypeVar("T", bound=WithValue)

class WithValueMixin(Generic[T]):
    pass

class C:
    @property
    def value(self) -> str | None:
        return None

class D(WithValueMixin[C]):
    pass
```
Basically, my `WithValueMixin` is a mixin class that is generic over "things with a `value`". It looks like "value" can no longer be defined as a property.

Note that this also fails with mypy, so I might need to change my code.

**Comment by alexprengere** [comment]
Note that the canonical change to make it work is to replace the protocol with this version

```python
class WithValue(Protocol):
    @property
    def value(self) -> str | None: ...
```
This will work on both `ty` (0.0.56 and 0.0.57), `mypy`, and will also work with concrete classes with a `value` defined as a normal attribute.

**Comment by carljm** [comment]
Yes, this was not a regression, but an intentional correctness fix. The protocol with `value: str | None` specifies a readable and writable attribute of that type. The class C with readable-only property `value` does not implement that protocol. All other type checkers agree that this is an error.

Using a property on the protocol is currently the correct way to specify that the protocol demands only readability. PEP 767 is an in-progress proposal to allow `ReadOnly` type qualifier to be used to specify this more explicitly and concisely.

Our diagnostic quality is not very good here, though -- we should a) add diagnostic context for an upper bound assignment failure (currently we give none), and b) even if we modify the example to just be a regular assignment (as below), our diagnostic context still bottoms out at "protocol member `value` is incompatible", ideally we'd explain why.

```py
from typing import Generic, Protocol, TypeVar

class WithValue(Protocol):
    value: str | None

class C:
    @property
    def value(self) -> str | None:
        return None
    
def f(wv: WithValue): ...

f(C())
```
